### PR TITLE
ENT-3249: Awood/fix profiles

### PR DESCRIPTION
**Pull request by awood**
## Testing:

Create an `acct_0` with 
```
bin/insert-mock-hosts --hbi --num-physical 10 --account acct_ --num-accounts 10 --clear
```

Opt in with
```
curl -X PUT -H "x-rh-identity: $(echo -n '{"identity":{"account_number":"acct_0","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w 0)" "http://localhost:8080/api/rhsm-subscriptions/v1/opt-in"
```

Run
```
curl -H "x-rh-identity: $(echo -n '{"identity":{"account_number":"acct_0","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"org123"}}}' | base64 -w 0)" "http://localhost:8080/api/rhsm-subscriptions/v1/capacity/products/OpenShift%20Container%20Platform?granularity=hourly&beginning=2019-06-19T00:00:00.000Z&ending=2019-06-19T23:59:59.999Z"
```

Before this PR, you'll get an error message about how HOURLY granularity isn't supported.  This was because "OpenShift Container Platform" wasn't found in the product profiles (since everything was indexed by engineering product ID) and accordingly a default profile was consulted.

With this fix, you'll get some results back.

### CODE REVIEW

File: src/main/java/org/candlepin/subscriptions/capacity/CapacityProductExtractor.java [hunk 1]
```diff
@@ -41,17 +41,17 @@
 public class CapacityProductExtractor {
 
     private static final Logger log = LoggerFactory.getLogger(CapacityProductExtractor.class);
-    private final Map<Integer, Set<String>> productIdToProductsMap;
+    private final Map<Integer, Set<String>> productToProductIdsMap;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/capacity/CapacityProductExtractor.java:44]:
```suggestion
    private final Map<Integer, Set<String>> engineeringProductIdToSubscriptionWatchProductsMap;
```
?

**Review by mstead — APPROVED** [review]
Agree that productId vs product is a little backwards. If it isn't already, we should create a tech debt card tracking the need to change this. I think this is something that we should do sooner rather than later.

Functionally, this addresses the issue that it was bumped back for.

@kahowell  can do the final merge once he is done his review.

File: src/main/java/org/candlepin/subscriptions/files/ProductProfile.java [hunk 2]
```diff
@@ -43,21 +43,21 @@ public static ProductProfile getDefault() {
...
     private String name;
-    private Set<SubscriptionWatchProductId> productIds;
+    private Set<SubscriptionWatchProduct> products;
     private Set<SyspurposeRole> syspurposeRoles;
     private Granularity finestGranularity;
     private boolean burstable = false;
     private String prometheusMetricName;
     private String prometheusCounterName;
-    private Map<String, String> architectureProductMap;
+    private Map<String, String> architectureProductIdMap;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfile.java:52]:
```suggestion
    private Map<String, String> architectureToSubscriptionWatchProductMap;
```
?

File: src/main/java/org/candlepin/subscriptions/files/SubscriptionWatchProduct.java [hunk 3]
```diff
@@ -24,24 +24,24 @@
 import java.util.Set;
 
 /** Represents the idea of products in Subscription Watch and what family they slot into. */
-public class SubscriptionWatchProductId {
-    private String id;
-    private Set<String> products;
+public class SubscriptionWatchProduct {
+    private String product;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/SubscriptionWatchProduct.java:28]:
```suggestion
    private String engineeringProductId;
```
?

File: src/main/java/org/candlepin/subscriptions/files/SyspurposeRole.java [hunk 4]
```diff
@@ -26,7 +26,7 @@
 /** Represents Syspurpose Roles and the product families they slot into. */
 public class SyspurposeRole {
     private String name;
-    private Set<String> products;
+    private Set<String> productIds;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/SyspurposeRole.java:29]:
```suggestion
    private Set<String> subscriptionWatchProductIds;
```
for clarity?

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 5]
```diff
@@ -33,89 +35,128 @@
 public class ProductProfileRegistry {
     private static final Logger log = LoggerFactory.getLogger(ProductProfileRegistry.class);
 
-    private final Map<String, ProductProfile> profileMap;
+    private final Map<Integer, ProductProfile> productToProfileMap;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:38]:
```suggestion
    private final Map<Integer, ProductProfile> engineeringProductIdToProfileMap;
```
?

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 6]
```diff
@@ -33,89 +35,128 @@
...
+    // NB: We should use ProductId as the key for type safety but that requires test updates
+    private final Map<String, ProductProfile> productIdToProfileMap;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:40]:
```suggestion
    private final Map<String, ProductProfile> SubscriptionWatchProductIdToProfileMap;
```
?

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 7]
```diff
@@ -33,89 +35,128 @@
...
+            throw new IllegalStateException("Failed to add profile " + profile.getName() +
+                ". The following productIds are already defined: " + duplicateIds);
+        }
+
+        profileProducts.stream()
+            .flatMap(x -> x.getProductIds().stream())
+            .forEach(x -> productIdToProfileMap.put(x, profile));
+    }
+
+    public ProductProfile findProfileForProductId(String productId) {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:93]:
```suggestion
    public ProductProfile findProfileForSubscriptionWatchProductId(String productId) {
```
?

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 8]
```diff
@@ -33,89 +35,128 @@
...
 
-        profileIds.forEach(x -> profileMap.put(x.getId(), profile));
+    public ProductProfile findProfileForProduct(String product) {
+        return findProfileForProduct(Integer.parseInt(product));
     }
 
-    public ProductProfile findProfile(String productId) {
-        if (profileMap.containsKey(productId)) {
-            return profileMap.get(productId);
+    public ProductProfile findProfileForProduct(Integer product) {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:109]:
```suggestion
    public ProductProfile findProfileForEngineeringProductId(Integer product) {
```
?

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 9]
```diff
@@ -33,89 +35,128 @@
...
+
+    public ProductProfile findProfileForProductId(String productId) {
+        if (productIdToProfileMap.containsKey(productId)) {
+            return productIdToProfileMap.get(productId);
         }
+        log.warn("ProductId {} not found in product profile registry. Returning default.", productId);
+        return ProductProfile.getDefault();
+    }
+
+    public ProductProfile findProfileForProductId(ProductId productId) {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:101]:
```suggestion
    public ProductProfile findProfileForSubscriptionWatchProductId(ProductId productId) {
```
?

File: src/main/java/org/candlepin/subscriptions/files/SubscriptionWatchProduct.java [hunk 10]
```diff
@@ -24,24 +24,24 @@
...
+    private Set<String> productIds;
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/SubscriptionWatchProduct.java:29]:
```suggestion
    private Set<String> subscriptionWatchProductIds;
```

File: src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java [hunk 11]
```diff
@@ -33,89 +35,128 @@
...
+        return findProfileForProductId(productId.toString());
+    }
 
-        profileIds.forEach(x -> profileMap.put(x.getId(), profile));
+    public ProductProfile findProfileForProduct(String product) {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/files/ProductProfileRegistry.java:105]:
```suggestion
    public ProductProfile findProfileForEngineeringProductId(String product) {
```

**Review by kahowell — CHANGES_REQUESTED** [review]
I found the changes a bit hard to follow. Apologies for the late input on naming, but seeing it here certainly helped.

I'd like to ask that we specifically call out in all method/field names which type of product we're talking about:

- engineeringProduct (e.g. 72)
- subscriptionWatchProduct (e.g. "RHEL Workstation")

I tried to identify the top-level field and method changes w/ suggestions (but possibly missed some).

**Review by kahowell — APPROVED** [review]
